memory/virtual_mem.cpp

```diff
@@ -124,20 +79,46 @@ std::vector<uint8_t> VirtualMem::LoadByteSequence(uintptr_t addr, uint64_t lengt
     return arr;
 }
 
-template <bool AllocPageIfNeeded>
-uint8_t *VirtualMem::GetPhysAddress(uintptr_t addr) const
+uint64_t VirtualMem::PageLookUp(uint32_t vpn0, uint32_t vpn1, uint32_t vpn2, uint32_t vpn3)
 {
-    uint64_t page_id = GetPageIdByAddress(addr);
-    uint64_t offset = GetPageOffsetByAddress(addr);
-    uint8_t *phys_addr = ram_->GetAddr<AllocPageIfNeeded>(page_id, offset);
-    assert(phys_addr != nullptr);
-    return phys_addr;
+    auto paddr3 = page_translation_[vpn3];
+    auto paddr2 = paddr3[vpn2];
+    auto paddr1 = paddr2[vpn1];
+    auto paddr0 = paddr1[vpn0];
+    if (paddr0 == 0) {
+        uint64_t pageNum = ram_->GetEmptyPageNumber();
+        ram_->InitPage(pageNum);
+        paddr0 = pageNum;
+        page_translation_[vpn3][vpn2][vpn1][vpn0] = pageNum;
```

Comment:
Maybe you can get reference to this earlier and don't do 4 times addresation in arrays again again?